Simplify hooks logic

wjhsf opened the pull request:
(This is a replay of #954 onto `develop`)

Over time, the implementations of query and mutation hooks have gotten a bit convoluted, and support has been added for features that we no longer deem necessary. This PR aims to simplify the signatures of the hooks and clean up the internal logic used.

As a result, query hooks now follow a pattern similar to:
```ts
const {
  data // Basket only, returning Response is no longer supported
} = useBasket(
  apiOptions // passed through to commerce-sdk-isomorphic
  queryOptions // passed through to @tanstack/react-query
)
``` 
Mutation hooks have also been simplified:
```ts
const { mutate } = useShopperBasketsMutation('createBasket') // Only takes the mutation name, now
mutate(apiOptions) // passed through to commerce-sdk-isomorphic
```

Under the hood, I've moved as much logic as I could from the endpoint hooks into the useQuery and useMutation helpers, to reduce repetition.

I've also refactored the query update logic. That must be implemented manually for each endpoint, so I've moved it out of the mutation.ts file, which is generated, and into a new config.ts file (per API, i.e. peer of mutation.ts). I've moved a lot of things around between the various locations (updateCache, useMutation, and the per-API methods), to hopefully simplify the interface as much as possible, and I've DRYed up some repeated logic.

In addition to cleaning up / fixing types, I made two major changes to query keys / cache updates: First, query keys previously only used parameters set on the options, but parameters can also be set on the config. So I've updated the logic to merge config/options before generating query keys / cache lookups. Secondly, we previously had cache lookup logic that accepted a query key, and created a lookup function from that. I've refactored the code so that we provide the lookup functions directly, which we can build from composable helpers. This provides us much more flexibility.

This PR will have a _lot_ of file changes, so to keep review a bit more focused, I opened a separate PR with a snapshot, https://github.com/SalesforceCommerceCloud/pwa-kit/pull/981.

<!--- Provide a short summary of your changes in the Title field above -->

# Description

<!--- A longer summary of your changes, including: a description of the issue that you’re addressing, a list of required dependencies (if applicable), and any other relevant context. -->

# Types of Changes

<!--- What types of changes does your code introduce? Put an `x` in all the boxes that apply: -->

- [ ] **Bug fix** (non-breaking change that fixes an issue)
- [ ] **New feature** (non-breaking change that adds functionality)
- [ ] **Documentation update**
- [ ] **Breaking change** (could cause existing functionality to not work as expected)
- [ ] **Other changes** (non-breaking changes that does not fit any of the above)

> Breaking changes include:
>
> - Removing a public function or component or prop
> - Adding a required argument to a function
> - Changing the data type of a function parameter or return value
> - Adding a new peer dependency to `package.json`

# Changes

- (change1)

# How to Test-Drive This PR

- (step1)

# Checklists

<!--- Enter an `x` in all the boxes that apply. -->
<!--- If you’re unsure about any of these, don’t hesitate to ask. We’re here to help! -->

## General

- [ ] Changes are covered by test cases
- [ ] CHANGELOG.md updated with a short description of changes (_not_ required for documentation updates)

## Accessibility Compliance

You must check off all items in **one** of the follow two lists:

- [ ] There are no changes to UI

_or..._

- [ ] Changes were tested with a Screen Reader (iOS VoiceOver or Android Talkback) and had no issues
- [ ] Changes comply with [WCAG 2.0 guidelines levels A and AA](https://www.wuhcag.com/wcag-checklist/)
- [ ] Changes to common UI patterns and interactions comply with [WAI-ARIA best practices](https://www.w3.org/TR/wai-aria-practices-1.1/)

## Localization

- [ ] Changes include a UI text update in the Retail React App (which requires translation)

wjhsf committed:
Merge hooks changes from wjh/simplify-hooks

wjhsf committed:
lint fixes

wjhsf committed:
Use correct type.

wjhsf committed:
Remove unnecessary file.

alexvuong commented on packages/commerce-sdk-react/src/hooks/ShopperBaskets/config.ts:
```diff
@@ -0,0 +1,79 @@
+    const arg = {basketId}
+    return basketId
+        ? {
+              update: [['/baskets', basketId, arg]]
```
Where is the update for customer baskets on basket mutations?

alexvuong commented on packages/commerce-sdk-react/src/hooks/Test/query.ts:
```diff
@@ -0,0 +1,234 @@
+            '.api.commercecloud.salesforce.com/checkout/shopper-baskets/',
+            parameters.version,
+            '/organizations/',
+            parameters.organizationId,
```
I assume here that we set those values as queryKey is because we want to cache different sets of commerce config. Would this be a bit overkill?

wjhsf replied:
It definitely is overkill; I wasn't sure what data we wanted to capture for query keys, so I started off by capturing it all. After a bit more exposure to the code, I'm thinking a good idea would be to have the filled-out template path and an object at the end with path + query parameters. (`['/baskets/', parameters.basketId, /* ..., */ parameters]`) I think that this will give us the flexibility we need (e.g. invalidate most `/baskets/{basketId}`, but selectively update those with matching `locale` query parameter so we don't get the wrong language somewhere) while not having too cumbersome a syntax (e.g. having to remember the order of the query parameters, as exists in the current state of this PR).

alexvuong commented on packages/commerce-sdk-react/src/hooks/utils.ts:
```diff
@@ -1,86 +1,64 @@
-        queryClient.setQueryData(queryKey, updater)
+    // STEP 1. Update data inside query cache for the matching queryKeys
+    cacheUpdates.update?.forEach((queryKey) => {
+        queryClient.setQueryData(queryKey, response)
```
Note: we've updated the signature for setQueryData to use a function `updater` because we want to update manually the cache on some mutations (e.g basket mutation) to

wjhsf committed:
Add support for updater in cache updates.

wjhsf committed:
Sort imports.

wjhsf committed:
Change `updater` generic from being the same for all to being per function.

wjhsf committed:
Clean up cache update matrix.

wjhsf committed:
Rekerjigger cache logic to see what we can do with it.

wjhsf committed:
Merge branch 'develop' into wjh/hooks-on-develop

wjhsf committed:
Merge branch 'develop' into wjh/hooks-on-develop

wjhsf committed:
Create options merge util.

wjhsf committed:
Use merged options in query hooks and simplify query keys.

wjhsf committed:
Simplify endMatches logic.

wjhsf committed:
Add predicate helper to match API config values.

wjhsf committed:
Fix temp test file.

wjhsf committed:
Use consistent order of generic parameters.

wjhsf committed:
Remove pointless line.

wjhsf committed:
Use consistent generic parameter names.

wjhsf committed:
Add comments.

wjhsf committed:
More comments.

wjhsf committed:
Remove test directory.

wjhsf committed:
Generate hooks for all APIs

wjhsf committed:
Update Shopper Baskets config from Test directory.

wjhsf committed:
Update Shopper Experience query hooks.

wjhsf committed:
Fix missed merged options type.

wjhsf committed:
Add Shopper Login hooks.

wjhsf committed:
Convert getGiftCertificate from mutation hook to query hook.

wjhsf committed:
Update Shopper Login hooks.

Change logoutCustomer from query to mutation.
Change useRetrieveCredQualityUserInfo to getCredQualityUserInfo.

wjhsf committed:
First pass at converting to new cache update matrix.

wjhsf committed:
Remove canceled Shopper Discovery Search API.

wjhsf committed:
Update login helpers.

wjhsf committed:
Fix auth types.

wjhsf committed:
Hopefully avoid caching getGiftCertificate.

wjhsf committed:
Rename shopper login helper to auth helper.

wjhsf committed:
Move auth helper to hooks root.

wjhsf committed:
Mark useAuthorizationHeader as internal.

wjhsf committed:
Rename useAuth to useAuthContext for clarity w/ useAuthHelper.

wjhsf committed:
Convert useAuthHelper to default export for consistency.

wjhsf committed:
Remove siteId from AuthData, as it is passed via config.

wjhsf committed:
Update comment with generated code.

wjhsf committed:
Merge branch 'develop' into wjh/hooks-on-develop

wjhsf committed:
Move headers to last parameter because we rarely use it.

wjhsf committed:
Update cache update matrices to new interface.

adamraya reviewed (commented):
I left some comments on the PR with a snapshot from this one opened to facilitate the review.

The comments are mainly about the query key and the changes to the cache.

Snapshot PR comments 👉 https://github.com/SalesforceCommerceCloud/pwa-kit/pull/981.

bendvc committed:
Refactore test project after api changes.

wjhsf committed:
Clean up cache update logic.

Hopefully makes it much easier to maintain.

wjhsf committed:
Update names for clarity.

wjhsf committed:
Revert test file to version from develop.

wjhsf committed:
Merge remote-tracking branch 'origin/origin/bendvc/hooks-on-develop' into wjh/hooks-on-develop

wjhsf committed:
Add reminder to revert.

wjhsf committed:
Remove unnecessary default export.

wjhsf committed:
Merge branch 'develop' into wjh/hooks-on-develop

wjhsf committed:
Cache update TODOs should return functions, not throw.

wjhsf committed:
Allow user to override authorization header.

wjhsf committed:
Remove old TODO.

wjhsf committed:
Remove ability to set custom query key.

Cache update logic requires knowing what query keys are used,
allowing arbitrary keys makes it impossible to know.

wjhsf committed:
Replace getQueryKey callback with just declaring query key.

wjhsf committed:
Clean up types.

wjhsf committed:
Rename config.ts to cache.ts

wjhsf committed:
Restore type checking test files.

wjhsf committed:
Suppress react query error logs during tests.

wjhsf committed:
Strip unused parameters before generating query key.

wjhsf committed:
Fix tpe errors.

wjhsf committed:
Partially fix type errors.

wjhsf committed:
Update Shopper Baskets cache logic to use narrower parameters.

wjhsf committed:
Test that all endpoints have hooks.

wjhsf committed:
Create new mutation tests for refactored code.

wjhsf committed:
Add test to validate all mutations have cache update logic.

wjhsf committed:
Alphabetize cache updates.

wjhsf committed:
Simplify mutation tests a bit.

wjhsf committed:
Fix nock adding duplicate request mocks.

wjhsf committed:
Make failing tests have more helpful failures.

Show the hook's error instead of "hook timed out" or nock mismatch.

wjhsf committed:
Fix tests failing due to missing parameters.

wjhsf committed:
Replace old mutation test file with new one.

alexvuong commented on packages/commerce-sdk-react/src/auth/index.test.ts:
```diff
@@ -85,7 +85,7 @@ describe('Auth', () => {
         const auth = new Auth(config)
 
-        const sample = {
+        const sample: Omit<AuthData, 'refresh_token'> & {refresh_token_guest?: string} = {
```
Can we create a type variable for this one plz?

alexvuong commented on packages/commerce-sdk-react/src/hooks/ShopperBaskets/cache.ts:
```diff
@@ -0,0 +1,157 @@
+): CacheUpdate => ({
+    // TODO: We only update the basket from the matching locale; we should also invalidate other locales
+    ...updateBasketQuery(customerId, parameters, response),
+    ...invalidateCustomerBasketsQuery(customerId, parameters)
```
We already manually update the baskets in `updateBasketQuery`. I don't think we need to invalidate customerBaskets.

wjhsf committed:
Remove unused helper type.

alexvuong commented on packages/commerce-sdk-react/src/hooks/ShopperBaskets/cache.ts:
```diff
@@ -0,0 +1,157 @@
+    }
+
+    // We can only update customer baskets if we have a customer!
+    if (!customerId) return {update: [basketUpdate]}
```
In which case the customerId will be null? We only update the cache in onSuccess callback, and the customerId is passed by calling useCustomerId, which seems to be defined at all time. ( No API is proceeded til authentication is successful in which customer_id is defined)

wjhsf replied:
The return type from `useCustomerId` is `string | null`, so we need this check. (If the return type is incorrect, that should be fixed in a follow up story.)

wjhsf committed:
Create hook success/error helpers.

alexvuong commented on packages/commerce-sdk-react/src/test-utils.tsx:
```diff
@@ -135,3 +140,31 @@ export const assertInvalidateQuery = (
+    return method.replace(/^./, (ltr) => `use${ltr.toUpperCase()}`)
+}
+
+export const expectAllEndpointsHaveHooks = (
```
Should we include list if not_implemented list here to use for other hook rather than just basket?

wjhsf replied:
So this might not be the most intuitive thing, but I changed how we detect "not implemented", and this test doesn't quite check that. We don't have a "not implemented" list any more. Instead, "implemented" means "_value_ is not `undefined` in the `cacheUpdateMatrix`". In addition to that, I made a `TODO` helper that logs "this method is not implemented" and returns `undefined`. Then, in the `cacheUpdateMatrix`, we put `[method]: TODO(method)`. This means that the method is still "not implemented", but it's more explicitly visible because it's explicitly present in the `cacheUpdateMatrix` (and logs a warning to console).

This test doesn't quite check for "not implemented". It only checks the _keys_ of the `cacheUpdateMatrix`, rather than the _values_. So, to this test, every `TODO` is considered implemented. That way we never need an explicit list of "not implemented". But, writing this out, I feel like the keys vs values distinction is a bit too subtle, so I'll change the test to check values and have an explicit list of "not implemented" in the assertion.

alexvuong commented on packages/commerce-sdk-react/src/hooks/ShopperBaskets/mut.test.ts:
```diff
@@ -0,0 +1,189 @@
+type BasketsResult = ShopperCustomersTypes.BasketsResult
+
+/** Create an options object for Shopper Baskets endpoints, with `basketId` pre-filled. */
+const makeOptions = <Method extends Exclude<keyof Client, 'clientConfig'>>(
```
Nit: `createOptions` sounds better :P

wjhsf replied:
Counterpoint, it's 2 more characters! :P

alexvuong commented on packages/commerce-sdk-react/src/hooks/ShopperBaskets/mut.test.ts:
```diff
@@ -0,0 +1,189 @@
+}
+const oldCustomerBaskets: BasketsResult = {
+    // We aren't implementing the full basket, so we assert to pretend we are
+    baskets: [{basketId: 'other_basket'}, oldBasket] as BasketsResult['baskets'],
```
We need to make sure that the basketId here is consistent with the basket response from mutation, if we re-use the oldBasket/newBasket data as a response for mutation, we have to make sure this basketId is 'BASKET_ID'. Otherwise, the mutation won't get updated since we check the id before we allow the mutation to update. 
I'd advise we use the same Id as oldBasket/newBasket to avoid any confusion since we are likely to just involve one basket for all tests

wjhsf replied:
What do you mean here? The basket being used _is_ `oldBasket` (and, `newBasket` for `newCustomerBaskets`), so the ID (and full basket contents) are re-used. There are two different baskets in the `baskets` array to validate that the correct basket gets updated.

wjhsf committed:
Improve parameter matching.

wjhsf committed:
Add `deleteBasket` tests.

wjhsf committed:
Add request mock to fix test.

wjhsf committed:
Add TODO cache update logic for missing mutations.

alexvuong commented on packages/commerce-sdk-react/src/hooks/ShopperBaskets/mutation.test.ts:
```diff
@@ -1,267 +1,224 @@
-        cases.forEach(({name, assertions}) => {
-            test(name, assertions)
-        })
+const deleteTestCase = ['deleteBasket', makeOptions<'deleteBasket'>(undefined, {})] as const
```
Aren't `remove...` where we call delete to the api counted as deleteTestCase? What makes this one different then the `removeItemFromBasket? 
I assume deleteTestCase does not return any data onSuccess? 🤔

wjhsf replied:
`removeItemFromBasket` and the other endpoints _modify_ the basket (by removing nested data), and they all return `Basket`. Only `deleteBasket` _deletes_ the basket, and you're correct, it returns `void` (since there's no `Basket` to return).

wjhsf committed:
Change onSuccess from bound to unbound function.

I'm not sure why, but testing mutations that throw an error in `onSuccess`
only works with unbound functions. When an unbound function is used, the
error is properly reported in the hook result. With a bound function, the
test framework throws an error.

wjhsf committed:
Add test for not implemented mutations.

wjhsf committed:
Pass not implemented tests.

wjhsf committed:
Remove unused import.

wjhsf committed:
Merge branch 'develop' into wjh/hooks-on-develop

wjhsf committed:
Implement `resetPassword` caching as a no-op.

wjhsf committed:
Add Shopper Baskets query tests.

wjhsf committed:
Add "all endpoints have hooks" tests for all APIs.

wjhsf committed:
Fix failing index tests.

wjhsf committed:
Implement tests for all query hooks.

wjhsf committed:
Remove hooks for `authorizeCustomer` and `getTrustedAgentAuthorizationToken`.

These endpoints modify headers, rather than mutate or return data, so they
don't make sense for either query or mutation hooks. (At least, for our
current implementation.)

wjhsf committed:
Update "not implemented" tests to check if cache update logic exists.

wjhsf committed:
Extract reused type into type def.

wjhsf committed:
Update comment to reflect changed tests.

wjhsf committed:
Remove unnecessary `async`.

wjhsf committed:
Implement Shopper Contexts mutation tests.

wjhsf committed:
Rename `makeOptions` to `createOptions`.

wjhsf committed:
Implement Shopper Login mutation tests.

wjhsf committed:
Add extra ResponseError assertion.

wjhsf committed:
Update test names.

wjhsf committed:
Implement Shopper Orders mutation tests.

wjhsf committed:
Convert TODO from throwing to just logging.

wjhsf committed:
Fix failing "all endpoints have hooks" tests.

wjhsf committed:
Remove unused imports.

wjhsf committed:
Introduce query key helpers.

wjhsf committed:
Remove no longer needed query test file.

All query hooks are now ttested in their respective folders.

wjhsf committed:
Merge branch 'develop' into wjh/hooks-on-develop

wjhsf committed:
Update hook usage.

wjhsf committed:
Implement basic Shopper Customers mutation tests.

wjhsf committed:
Export query key types.

wjhsf committed:
Remove old file.

kevinxh commented on packages/commerce-sdk-react/src/auth/index.ts:
```diff
@@ -29,24 +30,25 @@ interface JWTHeaders {
+/** A shopper could be guest or registered, so we store the refresh tokens individually. */
 type AuthDataKeys =
-    | 'access_token'
-    | 'customer_id'
-    | 'enc_user_id'
-    | 'expires_in'
-    | 'id_token'
-    | 'idp_access_token'
+    | Exclude<keyof AuthData, 'refresh_token'>
     | 'refresh_token_guest'
     | 'refresh_token_registered'
-    | 'token_type'
-    | 'usid'
-    | 'site_id'
-    | 'customer_type'
+
```
Nice! Cleaner!

kevinxh commented on packages/commerce-sdk-react/src/hooks/ShopperBaskets/cache.ts:
```diff
@@ -0,0 +1,186 @@
+    '/baskets/',
+    parameters.basketId
+]
+const getCustomerBasketsPath = (parameters: GetCustomerBasketsParameters) => [
+    ...getBasePath(parameters),
+    '/customers/',
+    parameters.customerId,
+    '/baskets' // No trailing / as it's an aggregate endpoint
+]
```
If we have the leading slash for every path segment, then we never need to worry about trailing slashes. 

```suggestion
const getCustomerBasketsPath = (parameters: GetCustomerBasketsParameters) => [
    ...getBasePath(parameters),
    '/customers',
    `/${parameters.customerId}`,
    '/baskets' // No trailing / as it's an aggregate endpoint
]
```

wjhsf replied:
The trade off there is that then we don't have array elements that are _just_ the parameters. Which isn't much of a trade off, so 🤷. In any case, the introduction of the query key helpers completely eliminates the need to ever even know what the query key actually is, so I'll mark this as resolved. With the helpers, changing the query key format becomes pretty trivial, so we can always revisit.

kevinxh commented on packages/commerce-sdk-react/src/hooks/ShopperBaskets/cache.ts:
```diff
@@ -0,0 +1,186 @@
+            return {
+                ...oldData,
+                // Shopper Customers and Shopper Baskets have different definitions for the `Basket`
+                // type. (99% similar, but that's not good enough for TypeScript.)
```
😢

johnboxall replied:
If yah'll can describe this problem to me exactly, we can consider raising it with the relevant teams to align the type definitions.

wjhsf replied:
We discussed in slack last week, I think it was. I raised the issue with Pratik, but said it's not urgent. You can follow up and raise the priority, though.

wjhsf committed:
Update cache logic to use query keys instead of predicates.

wjhsf committed:
Update parser for proper TypeScript support.

babel-eslint is deprecated and superseded by @babel/eslint-parser.

wjhsf committed:
Remove unused babel-eslint.

wjhsf committed:
Merge branch 'develop' into wjh/hooks-on-develop

wjhsf committed:
Temporarily(?) use internal-lib-build so TypeScript files are properly linted.

wjhsf committed:
Fix eslint and tsc errors.

wjhsf committed:
Temporarily suppress linter errors.

kevinxh commented on packages/commerce-sdk-react/src/hooks/ShopperBaskets/cache.ts:
```diff
@@ -0,0 +1,186 @@
+/** Parameters that get passed around, includes client config and possible parameters from other endpoints */
+type BasketParameters = MergedOptions<Client, Argument<Client['getBasket']>>['parameters']
+/** Parameters that we actually send to the API for `getBasket` */
+type GetBasketParameters = Argument<ShopperBaskets<{shortCode: string}>['getBasket']>['parameters']
```
why is shortCode so special here?

wjhsf replied:
It's a required parameter for all API clients. Can't do anything without it.

kevinxh commented on packages/commerce-sdk-react/src/hooks/ShopperContexts/query.ts:
```diff
@@ -1,20 +1,46 @@
- *
  * A hook for `ShopperContexts#getShopperContext`.
- * Gets the shopper's context based on the shopperJWT.
+ * Gets the shopper's context based on the shopperJWT. ******** This API is currently a work in progress, and not available to use yet. ********
```
The js comments like these are usually not updated often and certainly not in the process when we release API changes. I wonder if there is a better place for such notes. Just a nit worry that we might forget to remove it in the future.

kevinxh reviewed (approved):
The PR looks good to me, is it the intention to move the changes in `pwa-kit-dev` to another PR?

alexvuong commented on packages/commerce-sdk-react/src/hooks/types.ts:
```diff
@@ -1,21 +1,61 @@
+}
+
+/** Change string index type to `never`. */
+type StringIndexNever<T> = {
```
This makes it more readable
```suggestion
type StringIndexToNever<T> = {
```

wjhsf committed:
Implement tests for shopper customers mutations that modify a customer.

wjhsf committed:
Implement full Shopper Customers mutation tests.

wjhsf committed:
Merge branch 'develop' into wjh/hooks-on-develop

wjhsf committed:
Rename `StringIndexNever` to `StringIndexToNever`

wjhsf committed:
Remove unnecessary invalidation.

The query is already updated, so doesn't need to be invalidated.

wjhsf committed:
Update generated comments for hooks.

wjhsf committed:
Remove unused variable.

wjhsf committed:
Remove SLAS `authenticateCustomer` and `getPasswordResetToken` hooks.

They modify state and return headers, rather than body, so they are not
suitable for the current state of mutation hooks.

wjhsf committed:
Remove deprecated Shopper Customers mutations.

wjhsf committed:
Comment out Shopper Customers endpoints in closed beta.

wjhsf committed:
Fix Shopper Customers tests.

wjhsf committed:
Fix Shopper Login tests.

wjhsf committed:
Update "hook exists" tests to distinguish between "mutation in enum" and "mutation has cache logic"

wjhsf committed:
Merge branch 'develop' into wjh/hooks-on-develop
